src/commands.cpp

```diff
@@ -2003,60 +2017,211 @@ void CommandGetUA::procresult()
+        string tmpstr;
+//        bool nonHistoric = (attributename.at(1) == '!');
+
+        // handle the attribute data depending on the scope
+        char scope = attributename.at(0);
+        switch (scope)
         {
-            string d;
-            d.assign((char*)data, l);
+        case '*':   // private
 
-            // Is the data a multiple of the cipher blocksize, then we're using
-            // a zero IV.
-            if (l % client->key.BLOCKSIZE == 0)
+            // decrypt the data and build the TLV records
+            tlvRecords = TLVstore::containerToTLVrecords(string((char*) data, datalen), &client->key);
+            if (!tlvRecords)
             {
-                if (!PaddedCBC::decrypt(&d, &client->key))
+                LOG_err << "Cannot extract TLV records for private attribute " << attributename;
+                client->app->getua_result(API_EINTERNAL);
+                delete [] data;
+                return;
+            }
+
+            if (attributename == "*keyring")
+            {
+                // Ed25519 sanity checkup
+                if (tlvRecords->find(EdDSA::TLV_KEY))
                 {
-                    delete[] data;
-                    client->app->getua_result(API_EINTERNAL);
-                    return;
+                    tmpstr = tlvRecords->get(EdDSA::TLV_KEY);
+
+                    // check if received private key seed is different from existing one
+                    if (client->signkey && memcmp(client->signkey->keySeed, tmpstr.data(), EdDSA::SEED_KEY_LENGTH))
+                    {
+                        LOG_info << "Private key for Ed25519 doesn't match. Updating...";
+                        delete client->signkey; // discard existing keypair
+                        client->signkey = NULL;
+                    }
+
+                    if (!client->signkey)
+                    {
+                        client->signkey = new EdDSA((unsigned char *) tmpstr.data());
+                        signkeyUpdated = true;
+                    }
+                }
+                else    // create a new keypair
+                {
+                    if (client->signkey)
+                    {
+                        // this case should never happen: the SDK has already created the key, but not found in MEGA
+                        LOG_err << "Private key for Ed25519 not found, but it exists locally. Updating...";
+                        delete client->signkey; // discard existing keypair
+                        client->signkey = NULL;
+                    }
+
+                    client->signkey = new EdDSA();
+                    signkeyUpdated = true;
+                }
+
+                // x25519 sanity checkup
+                if (tlvRecords->find(ECDH::TLV_KEY))
+                {
+                    tmpstr = tlvRecords->get(ECDH::TLV_KEY);
+
+                    // check if received private key seed is different from existing one
```

Comment:
for x25519 it's not a "key seed" but really actually a "private key". But that's a technicality :-)